ocaml/xapi/xapi_psr.ml

```diff
@@ -0,0 +1,143 @@
+type r =
+  | Success
+  | Failed_during_accept_new_pool_secret
+  | Failed_during_send_new_pool_secret
+  | Failed_during_finish
+
+module type Impl = sig
+  type pool_secret
+
+  type pool_secrets = pool_secret * pool_secret (* (old, new) *)
+
+  type host
+
+  val save_checkpoint : string -> unit
+
+  val retrieve_checkpoint : unit -> string option
+
+  val backup : pool_secrets -> unit
+
+  val retrieve : unit -> pool_secrets
+
+  val tell_accept_new_pool_secret : pool_secrets -> host -> unit
+
+  val tell_send_new_pool_secret : pool_secrets -> host -> unit
+
+  val tell_cleanup_old_pool_secret : host -> unit
+end
+
+module Make =
+functor
+  (Impl : Impl)
+  ->
+  struct
+    open Impl
+
+    type checkpoint =
+      | No_checkpoint
+      | Backup
+      | Accept_new_pool_secret
+      | Send_new_pool_secret
+      | Finish
+
+    exception Cannot_parse_checkpoint of string
+
+    let string_of_checkpoint = function
```

Comment:
You can use the RPC derivers to generate serialization and deserialization functions for you, avoiding all the boilerplate.